data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ObjectWorker.java

```diff
@@ -50,24 +57,25 @@ void parseS3Object(final S3ObjectReference s3ObjectPointer) throws IOException {
                 .key(s3ObjectPointer.getKey())
                 .build();
 
-        final BufferAccumulator bufferAccumulator = BufferAccumulator.create(buffer, numberOfRecordsToAccumulate, bufferTimeout);
+        final BufferAccumulator<Record<Event>> bufferAccumulator = BufferAccumulator.create(buffer, numberOfRecordsToAccumulate, bufferTimeout);
 
         try (final ResponseInputStream<GetObjectResponse> object = s3Client.getObject(getObjectRequest)) {
             codec.parse(object, record -> {
                 try {
                     bufferAccumulator.add(record);
-                } catch (Exception e) {
+                } catch (final Exception e) {
                     LOG.error("Failed writing S3 objects to buffer.", e);
```

Comment:
I was going to record this metric, but the Buffer class already has a metric for the number of failures to write to a buffer.

The retry mechanism is something I was thinking about. Once concern I have is that this becomes a new buffer concept in Data Prepper. But, if it fully blocks the SQS queue, that could be appropriate. Another concern is that, at some point in time, the connection to S3 would close and the rest of the stream would be dropped.

Reply:
> Once concern I have is that this becomes a new buffer concept in Data Prepper. But, if it fully blocks the SQS queue, that could be appropriate.

I don't want to introduce another buffer concept in data prepper either. I like your proposal.

> Another concern is that, at some point in time, the connection to S3 would close and the rest of the stream would be dropped.

What is our current timeout? The AWS client provides us with a TCP keep alive option. Can we use that to prevent a closure?

Reply:
> What is our current timeout? The AWS client provides us with a TCP keep alive option. Can we use that to prevent a closure?

I expect it is the default of the AWS SDK. I've created #1500 to look into this.

Reply:
Perfect, thanks!